### ISSUE TITLE
libmints/matrix class error?

### ISSUE THREAD

**Issue by SebWouters** [psi4/psi4#274]
The following two snippets of code do not give the same result:
Wrong one:

    const int nao = wfn->aotoso()->rowspi( 0 );
    const int nmo = wfn->nmo();
    int * nmopi   = wfn->nmopi();

    SharedMatrix aotomo; aotomo = SharedMatrix( new Matrix( num_irreps, nao, nmopi ) );
    SharedMatrix work;     work = SharedMatrix( new Matrix( num_irreps, nao, nmopi ) );
    SharedMatrix AO_RDM; AO_RDM = SharedMatrix( new Matrix( nao, nao ) );

    aotomo->gemm( false, false, 1.0, wfn->aotoso(), Cmat,   0.0 );
      work->gemm( false, false, 1.0, aotomo,        MO_RDM, 0.0 );
    AO_RDM->gemm( false, true,  1.0, work,          aotomo, 0.0 );
    
Right one:

    int * nmopi = wfn->nmopi();
    int * nsopi = wfn->nsopi();
    const int nao = wfn->aotoso()->rowspi( 0 );

    SharedMatrix work;     work = SharedMatrix( new Matrix( "work",   num_irreps, nsopi, nmopi ) );
    SharedMatrix SO_RDM; SO_RDM = SharedMatrix( new Matrix( "SO RDM", num_irreps, nsopi, nsopi ) );
    SharedMatrix AO_RDM; AO_RDM = SharedMatrix( new Matrix( nao, nao ) );

      work->gemm( false, false, 1.0, Cmat, MO_RDM, 0.0 );
    SO_RDM->gemm( false, true,  1.0, work, Cmat,   0.0 );

    for ( int ao_row = 0; ao_row < nao; ao_row++ ){
       for ( int ao_col = 0; ao_col < nao; ao_col++ ){
          double value = 0.0;
          for ( int irrep = 0; irrep < num_irreps; irrep++ ){
             for ( int so_row = 0; so_row < nsopi[ irrep ]; so_row++ ){
                for ( int so_col = 0; so_col < nsopi[ irrep ]; so_col++ ){
                   value += SO_RDM->get( irrep, so_row, so_col ) * wfn->aotoso()->get( irrep, ao_row, so_row ) * wfn->aotoso()->get( irrep, ao_col, so_col );
                }
             }
          }
          AO_RDM->set( 0, ao_row, ao_col, value );
       }
    }

In the wrong one, it appears that there is no loop over the irreps. Only irrep 0 contributions are added to AO_RDM.

**Comment by dgasmith** [comment]
Hmm thats actually kind of interesting. So I think the misunderstanding here is when you make `Matrix( num_irreps, nao, nmopi )`, as we do not support different symmetries along different axises nao is replicated along the axis so you have a [nao, nao, …] by [nmopi[0], nmopi[1], …] matrix. So when you do the AOTOMO, MO_RDM, AOTOMO.T transformation what you should *should* end up with is a (nirrep, nao, nao) matrix where the ao contribution of each irrep is in a subsequent (nao, nao) block. However, since AO_RDM only has one dimension that GEMM will only collect the first dimension (irrep).

What you need to do is change:
`AO_RDM = SharedMatrix( new Matrix( nao, nao ) );`

to
`AO_RDM = SharedMatrix( new Matrix( num_irreps, nao, nao ) );`

and then sum through the irreps of AO_RDM. I don’t think we currently have a built-in routine to collapse irreps like this.

Im hard pressed to call this last part a bug since this follows through on the matrix algebra and automatically summing is probably not wise.

The first part where calling `gemm` on different irrep matrices should probably throw an error because this only works where the LHS has less irreps than the RHS; however, if the RHS has more irreps it will seg fault.

As I said before, this is an interesting error as its a strange mix of opaque code and a small bug. Hopefully this is both the truth and makes sense, coffee should kick in soon.

**Comment by dgasmith** [comment]
Hmm thats actually kind of interesting. So I think the misunderstanding here is when you make `Matrix( num_irreps, nao, nmopi )`, as we do not support different symmetries along different axises nao is replicated along the axis so you have a `([nao, nao, …]` by `[nmopi[0], nmopi[1], …])` matrix. So when you do the AOTOMO, MO_RDM, AOTOMO.T transformation what you should *should* end up with is a (nirrep, nao, nao) matrix where the ao contribution of each irrep is in a subsequent `(nao, nao)` block. However, since AO_RDM only has one dimension that `gemm` will only collect the first dimension (irrep).

What you need to do is change:
`AO_RDM = SharedMatrix( new Matrix( nao, nao ) );`

to
`AO_RDM = SharedMatrix( new Matrix( num_irreps, nao, nao ) );`

and then sum through the irreps of AO_RDM. I don’t think we currently have a built-in routine to collapse irreps like this.

Im hard pressed to call this last part a bug since this follows through on the matrix algebra and automatically summing is probably not wise.

The first part where calling `gemm` on different irrep matrices should probably throw an error because this only works where the LHS has less irreps than the RHS; however, if the RHS has more irreps it will seg fault.

As I said before, this is an interesting error as its a strange mix of opaque code and a small bug. Hopefully this is both the truth and makes sense, coffee should kick in soon.

**Comment by jturney** [comment]
Yes, this is a bug. The wrong one should produce an error.

Take a look at the ```remove_symmetry``` routine of ```Matrix```. Something like the following might work:

```
    MintsHelper helper(wfn->basisset(), wfn->options(), 0);
    SharedMatrix sotoao = helper.petite_list()->sotoao();

    SharedMatrix work;     work = SharedMatrix( new Matrix( "work",   num_irreps, nsopi, nmopi ) );
    SharedMatrix AO_RDM; AO_RDM = SharedMatrix( new Matrix( nao, nao ) );
    AO_RDM->remove_symmetry (work, sotoao);
```

There are various examples of using ```remove_symmetry``` in ```libmints/writer.cc``` and ```libscf_solver/hf.cc```.

**Comment by andysim** [comment]
Fixed by @jturney in PR #276.
